### ISSUE TITLE
lab 1

### ISSUE THREAD

**Issue by Varshith00** [Varshith00/aiml#1]
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, 
OneHotEncoder
from sklearn.cluster import KMeans
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)
data = {
    "Student_ID": range(1, 101),
    "Preferred_Subject": np.random.choice(["Machine Learning", "Data Science", "AI", "Data Analytics"], 100),
    "Learning_Style": np.random.choice(["Visual", "Auditory", "Kinesthetic"], 100),
    "Preferred_Language": np.random.choice(["Python", "R", "Java", "C++"], 100),
    "Study_Preference": np.random.choice(["Group Study", "Solo Study"], 100),
    "Likes_ML": np.random.choice([0, 1], 100)  # 1 if they like Machine Learning, else 0
}

df = pd.DataFrame(data)

encoder = LabelEncoder()
df['Preferred_Subject'] = encoder.fit_transform(df['Preferred_Subject'])
df['Learning_Style'] = encoder.fit_transform(df['Learning_Style'])
df['Preferred_Language'] = encoder.fit_transform(df['Preferred_Language'])
df['Study_Preference'] = encoder.fit_transform(df['Study_Preference'])

kmeans = KMeans(n_clusters=3, random_state=0)
df['Cluster'] = kmeans.fit_predict(df[['Preferred_Subject', 'Learning_Style', 'Preferred_Language', 'Study_Preference']])

plt.figure(figsize=(10, 6))
sns.scatterplot(data=df, x='Preferred_Subject', y='Learning_Style', hue='Cluster', palette='viridis')
plt.title("Clustering of Students Based on Interests")
plt.show()

X = df[['Preferred_Subject', 'Learning_Style', 'Preferred_Language', 'Study_Preference']]
y = df['Likes_ML']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

clf = DecisionTreeClassifier(random_state=42)
clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)
print("Decision Tree Accuracy:", accuracy_score(y_test, y_pred))

feature_importances = pd.Series(clf.feature_importances_, index=X.columns)
feature_importances.plot(kind='barh', color='skyblue')
plt.title("Feature Importance in Predicting 'Likes_ML'")
plt.show()

print("Feature Importances:\n", feature_importances)
